PM: dependencies: Add support for devices dependencies

ceolin opened the pull request:
- Resume all dependencies of device before resume the device
- Don't suspend a device if there are other devices depending on it

Commit:
scripts: gen_handles: Change output directory

Move dev_handles.c to misc/generate directory instead of let it hanging
in the build directory.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Add support for dependencies

Resume all dependencies before the device itself and do not
suspend (or change the device state) a device if there is any
other device depending on it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
shell: device: Print information about supported devices

Print all devices that depends on (or is supported) by a device. Just
like is done with required devices.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

JordanYates commented:
I don't believe that the "resume also resumes all dependents" approach is the correct path forward. At least if it is merged, I will need to revert it in my fork. Essentially the problem boils down to the fact that dependent devices do not need to be permanently active in order for a given device to operate, and can be very detrimental to overall power consumption.

Imagine an accelerometer operating in Wake-On-Motion mode, the device is definitely active, but by itself is only drawing a few uA. The accelerometer needs to be connected to a bus however, and from memory a nordic SPI bus consumes ~30uA while active. Assuming a 5uA base current draw, this change approximately quadruples the total power consumption of the device.

Power-aware drivers that I have implemented (currently closed-source unfortunately) follow the `spi_nor` approach of `acquire_device`/`release_device`, with the addition of turning on/off SPI buses in these functions:
https://github.com/zephyrproject-rtos/zephyr/blob/c94e53c9f5b6be29ce42b3cc8b50c6a27a487246/drivers/flash/spi_nor.c#L463-L497

gmarull commented:
> I don't believe that the "resume also resumes all dependents" approach is the correct path forward. At least if it is merged, I will need to revert it in my fork. Essentially the problem boils down to the fact that dependent devices do not need to be permanently active in order for a given device to operate, and can be very detrimental to overall power consumption.
> 
> Imagine an accelerometer operating in Wake-On-Motion mode, the device is definitely active, but by itself is only drawing a few uA. The accelerometer needs to be connected to a bus however, and from memory a nordic SPI bus consumes ~30uA while active. Assuming a 5uA base current draw, this change approximately quadruples the total power consumption of the device.
> 

That's a nice use case to take into account and that definitely can't be omitted. I think Linux solves this problem by letting users enable the "ignore children" flag in a device (it would be SPI in this case).

ceolin commented:
> > I don't believe that the "resume also resumes all dependents" approach is the correct path forward. At least if it is merged, I will need to revert it in my fork. Essentially the problem boils down to the fact that dependent devices do not need to be permanently active in order for a given device to operate, and can be very detrimental to overall power consumption.
> 
> Imagine an accelerometer operating in Wake-On-Motion mode, the device is definitely active, but by itself is only drawing a few uA. The accelerometer needs to be connected to a bus however, and from memory a nordic SPI bus consumes ~30uA while active. Assuming a 5uA base current draw, this change approximately quadruples the total power consumption of the device.
> 

That's a nice use case to take into account and that definitely can't be omitted. I think Linux solves this problem by letting users enable the "ignore children" flag in a device (it would be SPI in this case). 

I am affraid I have not understood it properly. The code is not resuming childreen, it is resuming parents. And unless there is an additional information, that is the only thing that the system can do. The system knows that device *A* depends on device *B*, so when resuming device *A* it has to resume device *B*. What this code is doing regarding childreen is, if the system is going to suspend device *B* it checks if any of its children are resumed (device *A* in this case). If so, it does not suspend *B*.

ceolin commented:
> > > I don't believe that the "resume also resumes all dependents" approach is the correct path forward. At least if it is merged, I will need to revert it in my fork. Essentially the problem boils down to the fact that dependent devices do not need to be permanently active in order for a given device to operate, and can be very detrimental to overall power consumption.
> > 
> > 
> > Imagine an accelerometer operating in Wake-On-Motion mode, the device is definitely active, but by itself is only drawing a few uA. The accelerometer needs to be connected to a bus however, and from memory a nordic SPI bus consumes ~30uA while active. Assuming a 5uA base current draw, this change approximately quadruples the total power consumption of the device.
> 
> > That's a nice use case to take into account and that definitely can't be omitted. I think Linux solves this problem by letting users enable the "ignore children" flag in a device (it would be SPI in this case).
> 
> I am affraid I have not understood it properly. The code is not resuming childreen, it is resuming parents. And unless there is an additional information, that is the only thing that the system can do. The system knows that device _A_ depends on device _B_, so when resuming device _A_ it has to resume device _B_. What this code is doing regarding childreen is, if the system is going to suspend device _B_ it checks if any of its children are resumed (device _A_ in this case). If so, it does not suspend _B_.

Btw, the `ignore_children` is a nice property to have. Going to add it.

nashif commented on include/device.h:
```diff
@@ -382,6 +382,13 @@ struct device {
+	 * depends on this node.  The individual sets are extracted
+	 * with dedicated API, such as device_supported_handles_get().
+	 */
+	const device_handle_t *const supported_handles;
```
is supported the right terminology here? is this supposed to mean "Dependent Handles"

ceolin replied:
"supports" is the nomenclature used in devicetree.

ceolin replied:
btw this field will go away. Just keeping it while https://github.com/zephyrproject-rtos/zephyr/pull/37836 is not merged.

Commit:
pm: device: Add ignore children property

When a device goes to a low power state (anything different from
PM_DEVICE_STATE_ACTIVE), it checks if any device that depends on
it is active. If so, the device is kept active.

This property, ignore-children, bypass this behavior. In other words,
when this property is set, the device does not check if any children
is active and suspend the device unconditionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Do not check for ignored children

Use the ignore-children property in pm_device_state_set.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> > I don't believe that the "resume also resumes all dependents" approach is the correct path forward. At least if it is merged, I will need to revert it in my fork. Essentially the problem boils down to the fact that dependent devices do not need to be permanently active in order for a given device to operate, and can be very detrimental to overall power consumption.
> > Imagine an accelerometer operating in Wake-On-Motion mode, the device is definitely active, but by itself is only drawing a few uA. The accelerometer needs to be connected to a bus however, and from memory a nordic SPI bus consumes ~30uA while active. Assuming a 5uA base current draw, this change approximately quadruples the total power consumption of the device.
> 
> That's a nice use case to take into account and that definitely can't be omitted. I think Linux solves this problem by letting users enable the "ignore children" flag in a device (it would be SPI in this case).
https://github.com/zephyrproject-rtos/zephyr/pull/37723/commits/be7e6163090b3e28de43e4353a745f9fd6c0a485

nashif reviewed (dismissed):
please add a test.

Commit:
tests: pm: Add device dependency test

Add a test to check if device dependencies is working properly.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
device: Fix device from handle function

This function was returning NULL for the last handle because of a
wrong comparison.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: s/PM_DEVICE_FLAGS/PM_DEVICE_FLAG

Keep flags names consistent.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull reviewed (commented):
Changes LGTM, just left a few comments w.r.t. docs and a case I think it's not handled.

JordanYates reviewed (changes requested):
> I am affraid I have not understood it properly. The code is not resuming childreen, it is resuming parents. And unless there is an additional information, that is the only thing that the system can do. The system knows that device A depends on device B, so when resuming device A it has to resume device B.

If the only thing you can do is detrimental, you can choose not to do it. It is not true that a devicetree dependency between A and B implies that B must be in active mode for A to be in active mode.

A device driver is in the best position to make decisions about which dependencies must be active for any given state of operation. What is the use case that this change is trying to address?

ceolin commented:
That is why drivers can use "ignore-children" property.  This will make the driver be suspended independent of its children.

ceolin commented:
> That is why drivers can use "ignore-children" property. This will make the driver be suspended independent of its children.

Elaborating a little more and trying to understand what is problem that these changes are bringing to your use case. Some problems this pr address:

-  If you have two devices A and B sharing the same bus. Imagine that device A does what you said and does no require the bus powered on in some situations and then decides to power off the bus. This is a problem, since B requires the bus active. This PR address this problem (there is pm device runtime as well to do this more properly tbh).
- The device must be aware of which bus , power gate, ... These things can change according with the application and platform. It brings a lot of complexity to drivers to deal with.

I think you are thinking in a very particular case you have while this is a generic solution to cover most common cases. Of course, the solution has be good enough to cover these particular cases, so, my question is what is missing here to cover the problem you are describing.

JordanYates commented:
> That is why drivers can use "ignore-children" property. This will make the driver be suspended independent of its children.

That is not the direction I am concerned about, it is the child powering on the parents under the `if(bringup)` condition that is problematic.

> If you have two devices A and B sharing the same bus. Imagine that device A does what you said and does no require the bus powered on in some situations and then decides to power off the bus. This is a problem, since B requires the bus active. This PR address this problem (there is pm device runtime as well to do this more properly tbh).

Is that not the entire purpose of pm device runtime? Why introduce a different solution here that not only doesn't solve the problem, but breaks the current solution by ignores the usage counts?

> The device must be aware of which bus , power gate, ... These things can change according with the application and platform. It brings a lot of complexity to drivers to deal with.

While the entire devicetree context changes with application and platform, the concerns of a device driver do not. It is either on a power gate or not, it is presumably on a bus, and it has some set of GPIOs. Bus and GPIO's are trivial to control in the driver:
```
device_pm_get(bus, NULL, NULL);
do_comms_things(bus);
device_pm_put(bus);
```
IMO, the best way to solve the power gate issue is via a new runtime API, `pm_power_get(dev)`, `pm_power_put(dev)`, for applications/drivers to use when they need a device powered. For devices not on a power gate, it immediately returns.
For devices on a power gate, bringup/add dependencies on the power domain chain, notifying child devices that were previously turned off with `pm_device_state_set(dev, PM_DEVICE_STATE_SUSPEND)`.

> my question is what is missing here to cover the problem you are describing.

The ability for drivers to turn on devicetree dependents only when that dependent is actually needed for operation of the device.
Requiring the driver itself to disable the depenedents after this PR turns them on is not really a solution, as doing this on busses will cause transient transitions on the lines.

ceolin commented:
> > That is why drivers can use "ignore-children" property. This will make the driver be suspended independent of its children.
> 
> That is not the direction I am concerned about, it is the child powering on the parents under the `if(bringup)` condition that is problematic.
> 
> > If you have two devices A and B sharing the same bus. Imagine that device A does what you said and does no require the bus powered on in some situations and then decides to power off the bus. This is a problem, since B requires the bus active. This PR address this problem (there is pm device runtime as well to do this more properly tbh).
> 
> Is that not the entire purpose of pm device runtime? Why introduce a different solution here that not only doesn't solve the problem, but breaks the current solution by ignoring the usage counts?

It does not break more than it is now. Anyone using this api directly will ignore the usage count ...

> 
> > The device must be aware of which bus , power gate, ... These things can change according with the application and platform. It brings a lot of complexity to drivers to deal with.
> 
> While the entire devicetree context changes with application and platform, the concerns of a device driver do not. It is either on a power gate or not, it is presumably on a bus, and it has some set of GPIOs. Bus and GPIO's are trivial to control in the driver:
> 
> ```
> device_pm_get(bus, NULL, NULL);
> do_comms_things(bus);
> device_pm_put(bus);
> ```
> 
> IMO, the best way to solve the power gate issue is via a new runtime API, `pm_power_get(dev)`, `pm_power_put(dev)`, for applications/drivers to use when they need a device powered. For devices not on a power gate, it immediately returns.
> For devices on a power gate, bringup/add dependencies on the power domain chain, notifying child devices that were previously turned off with `pm_device_state_set(dev, PM_DEVICE_STATE_SUSPEND)`.

that is the goal, devices aware of their dependencies and applications should use device_pm_runtime. This API is intend (or envisioned) to be use by the system power management. The problem is that in practice there are no devices doing it properly. So this should avoid lead the system to an inconsistent situation. Maybe the solution is promote device_runtime_* to be the only interface devices/applications should use. Make it work for the use case you mentioned is just the case of adding an additional property in the device. But before move forward, I'll review it. That is not an odd behavior and other OSs seemed to do this, but I think we need more discussion.

> > my question is what is missing here to cover the problem you are describing.
> 
> The ability for drivers to turn on devicetree dependents only when that dependent is actually needed for operation of the device.
> Requiring the driver itself to disable the depenedents after this PR turns them on is not really a solution, as doing this on busses will cause transient transitions on the lines.

JordanYates commented:
> It does not break more than it is now. Anyone using this api directly will ignore the usage count ...

Yes it does. This change is in `pm_device_state_set`, which is used by `pm_device_get`. Therefore if I call `pm_device_get` on my accelerometer, SPI/I2C buses will now be permanently on. This is not currently the case.

> that is the goal, devices aware of their dependencies and applications should use device_pm_runtime.

Which this change breaks

gmarull reviewed (changes requested):
I've reviewed again, and current changes will also have a negative impact on system PM. In that situation, it is not required to manage each device dependencies since the wakeup/suspend sequence is already executed in the right order.

tejlmand commented on tests/subsys/pm/device_dependencies/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2021 Intel Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
should be update to 3.10 now.
```suggestion
cmake_minimum_required(VERSION 3.20.0)
```

ceolin commented:
> I've reviewed again, and current changes will also have a negative impact on system PM. In that situation, it is not required to manage each device dependencies since the wakeup/suspend sequence is already executed in the right order.

Which negative impact ? As it is in order it should not have any impact. Please, be specific

gmarull commented:
> > I've reviewed again, and current changes will also have a negative impact on system PM. In that situation, it is not required to manage each device dependencies since the wakeup/suspend sequence is already executed in the right order.
> 
> Which negative impact ? As it is in order it should not have any impact. Please, be specific 

There is no need to take care of each device children/parent relationships when doing system power state transitions (devices are traverser in order). The current modifications will add unnecessary overhead.

nashif commented:
> There is no need to take care of each device children/parent relationships when doing system power state transitions (devices are traversed in order).

and what order is that exactly? who sets it?

gmarull commented:
> > There is no need to take care of each device children/parent relationships when doing system power state transitions (devices are traversed in order).
> 
> 
> 
> and what order is that exactly? who sets it?

Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children. In this PR `pm_device_state_set` is modified so that dependencies (in parent or child direction) are traversed for each device, something not needed for system PM transitions.

nashif commented:
> Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children.

funny, we spent years talking about how this need to be changed and how it should use information from DTS then we added all the support needed to make this happen and we have this working and agreed for all device dependencies we would use DTS and have been building up toward this and now you tell me that the linker provided data is good enough...

ceolin commented:
> > > I've reviewed again, and current changes will also have a negative impact on system PM. In that situation, it is not required to manage each device dependencies since the wakeup/suspend sequence is already executed in the right order.
> > 
> > Which negative impact ? As it is in order it should not have any impact. Please, be specific 
> 
> There is no need to take care of each device children/parent relationships when doing system power state transitions (devices are traversed in order). The current modifications will add unnecessary overhead.

I know how this is traversed ... My question was about the "negative" impact that obviously resumes to a function call ....  Btw these callbacks should  cover also dependencies that are not specified on DT. Which is not covered by linkage.....

ceolin commented:
> > > Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children.
> > 
> > 
> > 
> > funny, we spent years talking about how this need to be changed and how it should use information from DTS then we added all the support needed to make this happen and we have this working and agreed for all device dependencies we would use DTS and have been building up toward this and now you tell me that the linker provided data is good enough...
> 
> We already had this discussion. For system PM it is sufficient to have a sorted list, which we already have (not according to DT ordinals, though, so not ideal). For runtime PM we need the dependency chain per-device, a slightly different thing.

The linkage is not enough ... It does not cover dependencies for devices defined manually .... And this implementation was discussed and you had agreed  with it. Of course @JordanYates brought a new perspective but you are talking as if you have not agreed or looked it. What is a little bit annoying

gmarull commented:
> The linkage is not enough ... It does not cover dependencies for devices defined manually ....

As far as I know there is currently no way to define dependencies between manually defined devices, so I don't see what is the problem here.

 And this implementation was discussed and you had agreed  with it. 

I'm not opposing to the proposal, I'm just saying that I have found an issue on it that I missed on my first review pass.

Of course @JordanYates brought a new perspective but you are talking as if you have not agreed or looked it. What is a little bit annoying

I may be missing something here...?

nashif commented:
> I may be missing something here...?

what is it that you are requesting to be changed in this PR exactly? Maybe this is the confusing part.

> We already had this discussion. For system PM it is sufficient to have a sorted list, which we already have (not according to DT ordinals, though, so not ideal). For runtime PM we need the dependency chain per-device, a slightly different thing.

It is not a different thing. For PM we should use exactly the same order. Once we have this in places, we can also look into moving the boot order to depend on DTS, something that was also discussed but not immediately possible.

ceolin commented:
> 
> > The linkage is not enough ... It does not cover dependencies for devices defined manually ....
> 
> As far as I know there is currently no way to define dependencies between manually defined devices, so I don't see what is the problem here.
> 

The foreach callback is intended to support inserted dependencies. So the way we are consuming it should no change. 

> > And this implementation was discussed and you had agreed  with it. 
> 
> I'm not opposing to the proposal, I'm just saying that I have found an issue on it that I missed on my first review pass.
> 
> > Of course @JordanYates brought a new perspective but you are talking as if you have not agreed or looked it. What is a little bit annoying
> 
> I may be missing something here...?
>

gmarull commented:
To illustrate the problem:

- We have that dev B depends on dev A.
- During resume, device A comes first (it will appear first in the device list in a properly configured system).
- Then it comes B. Before resuming B, `pm_device_state_set` it will iterate first over its parents and try to wake them up (dev A). Since dev A will be already resumed, nothing will happen, so behavior will be correct. My only concern is that we'll be wasting cycles to do an operation that we can skip.

I hope it is clear now. I may be wrong with what I've illustrated, but I think I have to raise my concerns for discussion.

JordanYates commented:
Another problem that I would consider blocking is that this control is all one way. It automatically turns things on but never turns them off. Parents are only iterated over in `if (bringup)`, which is only true when moving to active state.

So not only is the SPI bus in my theoretical example always on when the accelerometer is running, it is actually permanently on after the first use. This is bad.

JordanYates commented:
>> Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children.

> funny, we spent years talking about how this need to be changed and how it should use information from DTS then we added all the support needed to make this happen and we have this working and agreed for all device dependencies we would use DTS and have been building up toward this and now you tell me that the linker provided data is good enough...

I consider this change to be unrelated to the priority problem. Device priorities should definitely come from devicetree in some form, as a device that is initialized before a required parent has no chance of initializing properly. That is a different question from whether dependent devices should be turning on all of their parents automatically, which is what this change is doing.

gmarull commented:
> >> Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children.
> 
> 
> 
> > funny, we spent years talking about how this need to be changed and how it should use information from DTS then we added all the support needed to make this happen and we have this working and agreed for all device dependencies we would use DTS and have been building up toward this and now you tell me that the linker provided data is good enough...
> 
> 
> 
> I consider this change to be unrelated to the priority problem. Device priorities should definitely come from devicetree in some form, as a device that is initialized before a required parent has no chance of initializing properly. That is a different question from whether dependent devices should be turning on all of their parents automatically (and permanently), which is what this change is doing.

Agreed, it is a separate problem. It was brought to the discussion to clarify how system suspends/resumes all devices when the system is suspended/resumed, and how the change in this PR is affecting this path.

ceolin commented:
> Another problem that I would consider blocking is that this control is all one way. It automatically turns things on but never turns them off. Parents are only iterated over in `if (bringup)`, which is only true when moving to active state.
> 
> So not only is the SPI bus in my theoretical example always on when the accelerometer is running, it is actually permanently on after the first use. This is bad.

Nope, the system iterates in reverse order, so the accelerometer will be turned off before SPI and consequently SPI will have no dependency active and will be turned off.

ceolin commented:
> To illustrate the problem:
> 
>     * We have that dev B depends on dev A.
> 
>     * During resume, device A comes first (it will appear first in the device list in a properly configured system).
> 
>     * Then it comes B. Before resuming B, `pm_device_state_set` it will iterate first over its parents and try to wake them up (dev A). Since dev A will be already resumed, nothing will happen, so behavior will be correct. My only concern is that we'll be wasting cycles to do an operation that we can skip.
> 
> 
> I hope it is clear now. I may be wrong with what I've illustrated, but I think I have to raise my concerns for discussion.

This was clear since the very beginning. The problem is when the dependencies are not solved in compile time. Then iterate over children is necessary. Of course the current code is generating this dependency from the device tree only, but the handles array is supposed to cover inserted dependencies (that is something that I'll look next)

ceolin commented:
> > > > Device list is sorted by the linker using priorities. This means that in a properly configured system (ie priorities well configured) a parent will always be suspended/resumed after/before its children.
> > 
> > 
> > > funny, we spent years talking about how this need to be changed and how it should use information from DTS then we added all the support needed to make this happen and we have this working and agreed for all device dependencies we would use DTS and have been building up toward this and now you tell me that the linker provided data is good enough...
> > 
> > 
> > I consider this change to be unrelated to the priority problem. Device priorities should definitely come from devicetree in some form, as a device that is initialized before a required parent has no chance of initializing properly. That is a different question from whether dependent devices should be turning on all of their parents automatically (and permanently), which is what this change is doing.
> 
> Agreed, it is a separate problem. It was brought to the discussion to clarify how system suspends/resumes all devices when the system is suspended/resumed, and how the change in this PR is affecting this path.

It is not affecting the behavior, it is indeed introducing an extra function call though. But this may be necessary anyway.

JordanYates commented:
> Nope, the system iterates in reverse order, so the accelerometer will be turned off before SPI and consequently SPI will have no dependency active and will be turned off.

What is "the system" in this case? If something external is already turning things off in the right order, why is the code in this function turning things on? Or rather, why is the same code not turning things on and off?

JordanYates commented:
A larger problem for me is that I have no idea what these PM PR's are working towards in terms of an API.
It is hard to review piecemeal changes without larger context.
The same changes can be good or bad ideas depending on what the end goal is.

ceolin commented:
> > Nope, the system iterates in reverse order, so the accelerometer will be turned off before SPI and consequently SPI will have no dependency active and will be turned off.
> 
> What is "the system" in this case? If something external is already turning things off in the right order, why is the code in this function turning things on? Or rather, why is the same code not turning things on and off?

"the system" refers to the kernel, actually the pm subsystem. Basically when the idle thread runs and checks which power state the soc should go and based on this trigger some actions on devices (low power, suspend, resume, ...) This code came from the problem that several places are directly setting device states (not using device runtime API) and this can led to problems with devices being suspended while another driver is still depending on it.

JordanYates commented:
> "the system" refers to the kernel, actually the pm subsystem. Basically when the idle thread runs and checks which power state the soc should go and based on this trigger some actions on devices (low power, suspend, resume, ...) This code came from the problem that several places are directly setting device states (not using device runtime API) and this can led to problems with devices being suspended while another driver is still depending on it.

Alright, but this code is also used by the device runtime API, where there is no such 'hypervisor' automatically switching states.
Isn't the actual problem that drivers are arbitrarily controlling other devices states without taking into account other users. Why can't the offending drivers use the runtime API which does reference counting?

gmarull commented:
To summarize my concern with this change:

The function `_pm_device` is called to suspend all devices when system goes into a low power mode. Such function iterates over all devices in order (using priorities, not according to DT, but sufficient on a well configured system). As you can see, it uses `pm_device_state_set()` to do the operation, which in turn, will, in some cases, iterate over given device dependencies. This step is not necessary, since the outer function (`_pm_device`) is already iterating over all devices. The solution would be to avoid the dependency handling in such point, since it is not necessary. A future improvement could be to replace `z_device_get_all_static` with a list of devices sorted according to ordinals, but that is a separate issue.

https://github.com/zephyrproject-rtos/zephyr/blob/0f63d1135c5b1f1c78e52304b2d818c1491b1918/subsys/pm/device.c#L23-L55

ceolin commented:
> To summarize my concern with this change:
> 
> The function `_pm_device` is called to suspend all devices when system goes into a low power mode. Such function iterates over all devices in order (using priorities, not according to DT, but sufficient on a well configured system). As you can see, it uses `pm_device_state_set()` to do the operation, which in turn, will, in some cases, iterate over given device dependencies. This step is not necessary, since the outer function (`_pm_device`) is already iterating over all devices. The solution would be to avoid the dependency handling in such point, since it is not necessary. A future improvement could be to replace `z_device_get_all_static` with a list of devices sorted according to ordinals, but that is a separate issue.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/0f63d1135c5b1f1c78e52304b2d818c1491b1918/subsys/pm/device.c#L23-L55

As I have  mentioned, that is correct (no question), but it is not perfect. Some points:
 - It is fragile because drivers have to be aware of their dependent / supported devices when setting the priority (thought this is a whole initialization problem to be fair)
 - Requires a second array of pointers of the size of devices
 - Limited priority control, just a few levels allowed (don't know if it is problem)

But the main problem is that this function `pm_device_state_set` is being used to do "runtime pm" ins several situations and the fact that this does properly handle dependencies case several problems. It clearly won't be fixed to 2.7 so lets iterate and elaborate a whole solution for this problem.

gmarull commented:
I think it's easy to fix the issue I mentioned: just create a version of `pm_device_state_set` that doesn't care about dependencies, so it can be used for cases like `_pm_device`, and another one that does (re-using code as needed). I think this is the safest approach. Runtime PM will likely require a few more improvements to make it the default recommendation. I think that until it is battle tested and is proven to be able to solve most real use cases, it's always better to provide some low level shortcuts.
